packages/oauth-js/src/client.ts

```diff
@@ -0,0 +1,361 @@
+  authState: AuthState; // current state of authentication
+  error: string; // the last error message
+
+  private storage: Storage;
+  private timer: number | null;
+
+  constructor(config: ClientConfig, options?: CookieOptions) {
+    if (!config.clientId) throw new Error("A clientId is required");
+    if (!config.domain) throw new Error("A domain is required");
+
+    this.authState = "INIT";
+    this.error = "";
+    this.storage = getStorageAPI(!!options?.useCookie);
+    this.timer = null;
+
+    const port = ["80", "443"].includes(window.location.port)
+      ? ""
+      : `:${window.location.port}`;
+
+    this.config = {
+      ...config,
+      callbackUri:
+        config.callbackUri ||
+        `${window.location.protocol}//${window.location.hostname}${port}`,
+    };
+    this.options = {
+      sessionKey: config.sessionKey || SESSION_DATA_KEY,
+      ...DEFAULT_COOKIE_OPTIONS,
+      ...options,
+    };
+  }
+
+  async init() {
+    if (this.authState !== "INIT") {
+      return;
+    }
+
+    const tokenExpire = getTokenExpire(this.options);
+
+    if (hasAuthParams()) {
+      await this.exchange();
+    } else if (!!tokenExpire) {
+      // has a token and token will expire soon, do refresh
+      if (isTokenExpiring(tokenExpire)) {
+        try {
+          await this.refresh();
+          this.startTokenWatch();
+        } catch {}
+      } else {
+        // has a validate token
+        // TODO: call introspect to validate for opaque tokens
+        this.authState = "AUTHENTICATED";
+      }
+    } else {
+      this.authState = "NOAUTH";
+    }
+  }
+
+  async login() {
+    const { verifier, challenge } = await createPkceChallenge();
+
+    const stateCode = generateBase58(32);
+    this.storage.setItem(STATE_DATA_KEY, stateCode);
+    this.storage.setItem(VERIFIER_DATA_KEY, verifier);
+    this.storage.setItem(
+      LAST_PATH_KEY,
+      `${location.pathname}${location.search}`
+    );
+
+    const port = ["80", "443"].includes(window.location.port)
+      ? ""
+      : `:${window.location.port}`;
+
+    const url = buildAuthorizeUrl(this.config.domain, {
+      clientId: this.config.clientId,
+      redirectUri: this.config.callbackUri || "",
+      state: stateCode,
+      challenge: challenge,
+      challengeMethod: "S256",
+      responseType: "code_challenge",
+      scope: this.config.scope,
+    });
+
+    window.location.assign(url);
+  }
+
+  async logout() {
+    const response = await this.post("logout", {
+      token: getRefreshToken(this.options),
+      token_type_hint: "refresh_token",
+    });
+
+    if (response.ok) {
+      this.authState = "INIT";
+    } else {
+      this.authState = "ERROR";
+    }
+    removeTokenCookies(this.options);
+    this.storage.removeItem(STATE_DATA_KEY);
+  }
+
+  async refresh() {
+    const initialState = this.authState;
+    const lock = new Lock();
+
+    try {
+      this.authState = "UPDATING";
+
+      if (await lock.acquireLock(REFRESH_LOCK_KEY)) {
```

Comment:
As I read maybe this question gets answered, but just wondering if two tabs try to refresh, and the second one waits for the first to release the lock, it seems like this would then just refresh again? once the lock is released. Should there be a step after acquiring to determine if refresh is still needed?

Reply:
Yes, you're right, there needs to be a check after the lock is acquired. Will also add a forceRefresh option to override the check.

Reply:
added check